gapis/resolve/dependencygraph/footprint.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/google/gapid/gapis/capture"
+	"github.com/google/gapid/gapis/database"
+)
+
+var footprintBuildCounter = benchmark.GlobalCounters.Duration("footprint.build")
+
+// Footprint contains a list of command and a list of Behaviours which
+// describes the side effect of executing the commands in that list.
+type Footprint struct {
+	Commands                                []api.Cmd
+	Behaviours                              []*Behaviour
+	commandIndexToBehaviourIndexLookupTable *api.SubCmdIdxTrie
+	BehaviourIndices                        map[*Behaviour]uint64
+}
+
+// NewEmptyFootprint creates a new Footprint with an empty command list, and
+// returns a pointer to that Footprint.
+func NewEmptyFootprint(ctx context.Context) *Footprint {
+	return &Footprint{
+		Commands:                                []api.Cmd{},
+		Behaviours:                              []*Behaviour{},
+		commandIndexToBehaviourIndexLookupTable: &api.SubCmdIdxTrie{},
+		BehaviourIndices:                        map[*Behaviour]uint64{},
+	}
+}
+
+// NewFootprint takes a list of commands and creates a new Footprint with
+// that list of commands, and returns a pointer to that Footprint.
+func NewFootprint(ctx context.Context, cmds []api.Cmd) *Footprint {
+	return &Footprint{
+		Commands:                                cmds,
+		Behaviours:                              make([]*Behaviour, 0, len(cmds)),
+		commandIndexToBehaviourIndexLookupTable: &api.SubCmdIdxTrie{},
+		BehaviourIndices:                        map[*Behaviour]uint64{},
+	}
+}
+
+// Behaviour contains a set of read and write operations as side effect of
+// executing the command to whom it belongs. Behaviour also contains a refernce
+// to the back-propagation machine which should be used to process the
+// Behaviour to determine its liveness for dead code elimination.
+type Behaviour struct {
+	Reads    []DefUseAtom
+	Writes   []DefUseAtom
+	BelongTo api.SubCmdIdx
+	Alive    bool
+	Aborted  bool
+	Machine  BackPropagationMachine
+}
+
+// NewBehaviour creates a new Behaviour which belongs to the command indexed by
+// the given SubCmdIdx and shall be process by the given back-propagation
+// machine. Returns a pointer to the created Behaviour.
+func NewBehaviour(fullCommandIndex api.SubCmdIdx, m BackPropagationMachine) *Behaviour {
+	return &Behaviour{
+		BelongTo: fullCommandIndex,
+		Machine:  m,
+	}
+}
+
+// Read records a read operation of the given DefUseAtom to the Behaviour
+func (b *Behaviour) Read(c DefUseAtom) {
+	b.Reads = append(b.Reads, c)
+}
+
+// Write records a write operation of the given DefUseAtom to the Behaviour
+func (b *Behaviour) Write(c DefUseAtom) {
+	b.Writes = append(b.Writes, c)
+}
+
+// Modify records a read and a write operation of the given DefUseAtom to the
+// Behaviour
+func (b *Behaviour) Modify(c DefUseAtom) {
+	b.Read(c)
+	b.Write(c)
+}
+
+// DefUseAtom represents a piece of state which can be altered as commands
+// being executed.
+type DefUseAtom interface {
```

Comment:
I changed my mind, it will be renamed to DefUseVariable to be more consistent with the Wiki page: https://en.wikipedia.org/wiki/Live_variable_analysis